Fix behaviour of christioffel symbol functions and related Jacobians

kellertuer opened the pull request:
@jlumpe pointed out that

```
using Manifolds
M = MetricManifold(Euclidean(3), EuclideanMetric())
christoffel_symbols_first(M, zeros(3))
```
Throws an error, that `christoffel_symbols_first not implemented on Euclidean(3; field = ℝ)...`.

I was surprised by that and dived a little bit into our dispatch system for Christoffel symbols and their Jacobians.
I think I found the culprit.

In the original code the functions `christoffel_symbols_first`, `christoffel_symbols_second`,  `christoffel_symbols_second_jacobian` and `local_metric_jacobian` where introduced as transparent functions in our decorator system. This meant the following

1. the functions acted completely transparent when called with another decorator (group manifold for example)
2. the original implementation works for non-default metrics
3. the functions acted transparent for default metrics, i.e. falling back to calling `christoffel-symbols_first(M.manifold,...)` for those.

As far as I could see, the default implementation relied on functions that themselves might be transparent or not (`local_metric`, `local_metric_jacobian`,...) but was just a proper implementation for arbitrary metric manifolds anyways. This default implementation did not work for default metrics (as the one above).

So I changed the behaviour to the following: The function is now defined for an arbitrary `MetricManifold`, that is transparent for all others (1. is still valid), but the generic implementation is now called for 2. and 3.
The inner calls are now also working properly (falling back to `local_metric_jacobian` for `Euclidean` when using `MetricManifold(Euclidean(3), EuclideanMetric(3))`.

Now the above example works.

The one point left to discuss is:
The Christoffel functions now still have to be called with `MatricManifold`s, while they would be also valid for `Manifolds` (if one has a default metric in mind). One could try to do something like passing a `::Manifold` based call to the default metric `MetricManifold`. What do you think?

kellertuer commented:
I think there might even be more functions having the same problem, I will check them soon.

mateuszbaran commented:
Thanks for fixing it :+1: . 

> The one point left to discuss is:
> The Christoffel functions now still have to be called with `MatricManifold`s, while they would be also valid for `Manifolds` (if one has a default metric in mind). One could try to do something like passing a `::Manifold` based call to the default metric `MetricManifold`. What do you think?

I don't remember why we restricted metric-related functions to `MetricManifold`s but they definitely make sense for any manifold. What do you mean by default metric `MetricManifold` here? I don't think I understand your idea.

kellertuer commented:
Oh I meant the default metric `Manifold` that implicitly acts as a `MetricManifold` with default metric; for example for exp/log.

kellertuer commented:
The current state does currently not yet work, but my brain also stopped – this decorator stuff is quite challenging.

Missing:
* [ ] make metric functions available for manifolds with their implicit metric
* [ ] check documentation to be complete to this new extension

kellertuer commented:
While working on this I noticed that most probably all these metric functions lack a basis (that we might have implicitly assumed?), since when you work with the g_ij (elements of the metric matrix G) or the g^ij (its inverse) they are always with respect to a basis and hence the \Gamma_ij^k are, too?

kellertuer commented:
So. All docs are correct again, I am a little too lazy to add formal to Einstein and curvature tensor, but that would be a nice feature in the future I think.

mateuszbaran commented on src/manifolds/MetricManifold.jl:
```diff
@@ -1,8 +1,15 @@
+
+    Metric(M)
+
+Generate the manifold `M` with the correstponding metric (subtype of this type).
```
```suggestion
Generate the manifold `M` with the corresponding metric (subtype of this type).
```
Do you mean `MetricManifold` by "this type" or something else?

kellertuer replied:
Good that you ask, no, I meant the `Metric`. I wanted to clarify that there is `Metric(M)` for `MetricManifold(M,Metric())` as a constructor.

mateuszbaran replied:
Then the documentation is quite unclear, and this isn't exactly a constructor of `Metric` type since it returns a value of type `MetricManifold`. It should be more clearly stated I think. First, I'd add that it required a concrete subtype of `Metric` in the call, a second the description could be a bit more like:
```
    (metric::Metric)(M::Manifold)

Generate the `MetricManifold` that wraps the manifold `M` with given `metric`.
```

mateuszbaran replied:
Or similarly for the other method, `(::Type{T})(M::AbstractManifold) where {T<:AbstractMetric}`.

kellertuer replied:
Ah yes, that is a better signature. By the way, just for completeness we could also introduce `(M::Manifold)(m::Metric)` to do the same (which even feels more natural to me).

mateuszbaran replied:
> By the way, just for completeness we could also introduce `(M::Manifold)(m::Metric)` to do the same (which even feels more natural to me).

I don't have any opinions on that but if you like it then sure :slightly_smiling_face: .

kellertuer replied:
Maybe in a later PR to not make this PR too chaotic.

mateuszbaran commented:
Could you add some test that check that these methods work for, for example, `MetricManifold(Euclidean(3), EuclideanMetric())`?

kellertuer commented:
> Could you add some test that check that these methods work for, for example, `MetricManifold(Euclidean(3), EuclideanMetric())`?

Which “these methods” do you mean? All of them? (at first glance: christoffel_first, &_jacobian, christoffel_second, &_jacobian, Einstein, Ricci, local metric, inverse local metric, der local metric, inverse det local metric)?

mateuszbaran commented:
> Which “these methods” do you mean? All of them? (at first glance: christoffel_first, &_jacobian, christoffel_second, &_jacobian, Einstein, Ricci, local metric, inverse local metric, der local metric, inverse det local metric)?

Yes, exactly these, plus `riemann_tensor`.

kellertuer commented:
Ok, I will check for those in the evening (hopefully).

kellertuer commented:
It was a great idea to write the tests

1) I added `local_metric` and `inverse_local_metric` for `Euclidean` since now we have the proper transparency that passes the default metric (`EuclideanMetric`) down to the manifold itself. That's what we introduced the default-mechanism for.

2) noticed one bug in transparency

What's cool is that with just the local metric and its inverse all other functions worked :)

kellertuer commented:
I am surprised the codecov patch fails, since overall the code coverage increases. All new functionality (i.e. having the possibility of using all functions with an implicit default metric without the decorator) has tests so this PR is ready to merge.

mateuszbaran commented:
That coverage failure looks like a false positive.
